### DISCUSSION TITLE
Reproducibility

### DISCUSSION THREAD

**Discussion by do-me** [reinterpretcat/vrp#116] [Q&A] [answered]
I was wondering if reproducibility of results is an active **feature** or a more of a **side-effect** of VRP. 

I ran some very quick tests with the very simple problems with the WebAssembly version and it seems indeed to provide the same results. I couldn't find anything about it in the docs, so I thought I'd ask here. 

Am I correct to assume that: 
- as long as the input remains exactly the same and
- any termination criterion hits before the time limit is exceeded 

the results should always be the same or is there some kind of random seed somewhere for more complex problems?

**Comment by reinterpretcat** [comment, marked as answer]
No, algorithm is non-determenistic, so for different runs it can explore solution space differently and return different solutions. Search stability is one of the main topics among with solution quality and good termination criteria.

**Comment by do-me** [comment]
You're referring to this poll right? #82 
This is good to know, thank you!

**Reply by reinterpretcat** [reply]
Yes, it is mentioned here too.
At the moment, I'm lazily researching how to improve my current algorithm to address these issues rather working on new features.

**Reply by shubham441996** [reply]
Hey,
@reinterpretcat  Do we have any way to achieve this for now, or can you guide i am really interested in working on this

**Reply by reinterpretcat** [reply]
Yes, as I learned it is possible, but it would require some changes to pass `rng` object along the methods. Currently, it is accessed from thread local storage and it cannot be determenistic in multi-thread environment.

**Reply by shubham441996** [reply]
Thanks for the quick response @reinterpretcat . 
These are the following questions I have before diving deeper into the code.

- Using a static heuristic will make the solution less non-deterministic with parallelism 1?
- Can we achieve this via dynamic heuristics with parallelism?
- Will making it deterministic decrease the chances of getting the optimised outcome as we have in non-deterministic?

**Reply by reinterpretcat** [reply]
1. Hard to say: static selective heuristic just uses hard coded probabilities of heuristic to be called. Dynamic selective tries to learn them over search progression using some kind of variation Thompson Sampling. If you really need same output across different runs, better to change approach to fix non-determenism when same randomization seed is used.
2. Yes, it is achievable, but would require internal changes: mostly to pass rng object over internal api to avoid getting it from thread local storage. As rng object is supposed to be mutable without locking, the change will touch many methods.
3. It shouldn't. It will be still possible to get different results (more or less optimized) with different seeds.

**Reply by shubham441996** [reply]
I wanted to ask if deterministic behaviour is on your roadmap. If not, I’d be glad to take a stab at implementing it. Rust isn’t my primary language, but I’m happy to learn as I go.
